File: Sandbox/Code/Core/CameraController.cpp
#include "CameraController.h"

#include <algorithm>

void CameraController::OnUpdate()
{
    Input& input = GetContext().GetInput();
    Transform& cameraTransform = GetComponent<TransformComponent>().transform;

    // Look: only while holding the right mouse button. Lock/hide the cursor for the duration so
    // it can't run out of window to move in - SetCursorLockState hides it and, on Qt, warps it
    // back to center each move (SDL's relative mouse mode does the equivalent natively).
    if (input.GetMouseButtonDown(Input::MouseButton::Right))
        input.SetCursorLockState(Input::CursorLockState::Locked);
    else if (input.GetMouseButtonUp(Input::MouseButton::Right))
        input.SetCursorLockState(Input::CursorLockState::None);

    if (input.GetMouseButton(Input::MouseButton::Right))
    {
        constexpr float mouseSensitivity = 0.15f;

        Vector2Int mouseDelta = input.GetAxis(Input::AxisType::Mouse);

        m_Yaw += -static_cast<float>(mouseDelta.x) * mouseSensitivity;
        m_Pitch += -static_cast<float>(mouseDelta.y) * mouseSensitivity;
        m_Pitch = std::clamp(m_Pitch, -89.0f, 89.0f);

        // Rebuilt fresh from the total yaw/pitch every time, not composed incrementally - yaw
        // around world up, pitch around the yawed frame's local right, matching the classic
        // FPS-camera R = Yaw * Pitch order (see the m_Yaw/m_Pitch comment in CameraController.h
        // for why).
        cameraTransform.SetRotation(AngleAxis(Radians(m_Yaw), Vector3(0.0f, 1.0f, 0.0f)) * AngleAxis(Radians(m_Pitch), Vector3(1.0f, 0.0f, 0.0f)));
    }

    // Move: WASD + Space/Ctrl, relative to the camera's own orientation.
    constexpr float moveSpeed = 2.0f;

    Vector3 movement(0.0f);

    if (input.GetKey(KeyCode::W))
        movement += cameraTransform.GetForward();
    if (input.GetKey(KeyCode::S))
        movement -= cameraTransform.GetForward();
    if (input.GetKey(KeyCode::D))
        movement += cameraTransform.GetRight();
    if (input.GetKey(KeyCode::A))
        movement -= cameraTransform.GetRight();
    if (input.GetKey(KeyCode::SPACE))
        movement += Vector3(0.0f, 1.0f, 0.0f);
    if (input.GetKey(KeyCode::LCTRL))
        movement -= Vector3(0.0f, 1.0f, 0.0f);

    if (movement != Vector3(0.0f))
        cameraTransform.Translate(Normalize(movement) * moveSpeed * GetContext().GetTime().GetDeltaTime());
}

File: Sandbox/Assets/Scripts/camera_controller.py
import matcha_engine

# Port of Sandbox/Code/Core/CameraController. Free-fly movement (WASD + Space/Ctrl) and
# mouse-look (held right mouse button), relative to the camera entity's own orientation.

MOUSE_SENSITIVITY = 0.15
MOVE_SPEED = 2.0


class CameraController:
    # See RotationComponent's own copy of this comment for why these are bare annotations.
    entity: matcha_engine.Entity
    input: matcha_engine.Input
    time: matcha_engine.Time

    def __init__(self):
        # Accumulated separately (degrees) and rebuilt into the camera's rotation from scratch
        # each time, rather than incrementally composing rotations onto the running quaternion
        # frame after frame - see the same field in the original CameraController.h for why
        # (composing accumulates roll drift; recomputing yaw*pitch fresh from the totals doesn't).
        self.yaw = 0.0
        self.pitch = 0.0

    def on_update(self):
        transform = self.entity.get_transform()
        input = self.input

        # Look: only while holding the right mouse button. Lock/hide the cursor for the duration.
        if input.get_mouse_button_down(matcha_engine.MouseButton.RIGHT):
            input.set_cursor_lock_state(matcha_engine.CursorLockState.LOCKED)
        elif input.get_mouse_button_up(matcha_engine.MouseButton.RIGHT):
            input.set_cursor_lock_state(matcha_engine.CursorLockState.NONE)

        if input.get_mouse_button(matcha_engine.MouseButton.RIGHT):
            mouse_delta = input.get_axis(matcha_engine.AxisType.MOUSE)

            self.yaw += -float(mouse_delta.x) * MOUSE_SENSITIVITY
            self.pitch += -float(mouse_delta.y) * MOUSE_SENSITIVITY
            self.pitch = max(-89.0, min(89.0, self.pitch))

            # Yaw around world up, pitch around the yawed frame's local right - classic FPS-camera
            # R = Yaw * Pitch order.
            yaw_rotation = matcha_engine.angle_axis(matcha_engine.radians(self.yaw), matcha_engine.Vector3(0.0, 1.0, 0.0))
            pitch_rotation = matcha_engine.angle_axis(matcha_engine.radians(self.pitch), matcha_engine.Vector3(1.0, 0.0, 0.0))
            transform.set_rotation(yaw_rotation * pitch_rotation)

        # Move: WASD + Space/Ctrl, relative to the camera's own orientation.
        movement = matcha_engine.Vector3(0.0)

        if input.get_key(matcha_engine.KeyCode.W):
            movement = movement + transform.get_forward()
        if input.get_key(matcha_engine.KeyCode.S):
            movement = movement - transform.get_forward()
        if input.get_key(matcha_engine.KeyCode.D):
            movement = movement + transform.get_right()
        if input.get_key(matcha_engine.KeyCode.A):
            movement = movement - transform.get_right()
        if input.get_key(matcha_engine.KeyCode.SPACE):
            movement = movement + matcha_engine.Vector3(0.0, 1.0, 0.0)
        if input.get_key(matcha_engine.KeyCode.LCTRL):
            movement = movement - matcha_engine.Vector3(0.0, 1.0, 0.0)

        if movement != matcha_engine.Vector3(0.0):
            delta_time = self.time.get_delta_time()
            transform.translate(matcha_engine.normalize(movement) * MOVE_SPEED * delta_time)
